### PR TITLE
Adds Clang support for compiling PX4 on NuttX

### PR DESCRIPTION
**Pull request by mayanez**
Compiles PX4 codebase using Clang & arm-none-eabi tools.

Related Issue https://github.com/PX4/Firmware/issues/4840

I appreciate any feedback on the CMake files as I'm not that experienced with them. Specifically on how to take the include paths that I currently have hardcoded and retrieve them from the environment.

### CODE REVIEW

File: cmake/nuttx/px4_impl_nuttx.cmake [hunk 1]
```diff
@@ -530,6 +534,19 @@ function(px4_os_add_flags)
 	set(added_link_dirs
 		${nuttx_export_dir}/libs
 		)
+
+	if (${config_nuttx_hw} STREQUAL "m4")
+		set(APPEND added_link_dirs
+			/home/miguel/gcc-arm-none-eabi-5_4-2016q2/lib/gcc/arm-none-eabi/5.4.1/armv7e-m/fpu/
```
Comment (dagar) [inline cmake/nuttx/px4_impl_nuttx.cmake]: Hardcoded paths?

File: cmake/common/px4_base.cmake [hunk 2]
```diff
@@ -835,10 +855,23 @@ function(px4_add_common_flags)
...
+				-fomit-frame-pointer
+			)
+
+			list(APPEND cxx_warnings
+				-Wno-overloaded-virtual
+			)
+		else()
+			# force color for clang (needed for clang + ccache)
+			list(APPEND _optimization_flags
+				-fcolor-diagnostics
```
Comment (dagar) [inline cmake/common/px4_base.cmake]: Can this flag stay on? It makes development with clang much nicer.

File: cmake/nuttx/px4_impl_nuttx.cmake [hunk 3]
```diff
@@ -575,17 +611,78 @@ function(px4_os_add_flags)
...
+
+			if (${CMAKE_CXX_COMPILER_ID} MATCHES ".*Clang.*")
+				list(APPEND cpu_flags
+					-m32
+					-target arm-none-eabi
+					-fno-builtin
+				)
+
+				set(clang_cxx_flags
+					-isystem /home/miguel/gcc-arm-none-eabi-5_4-2016q2/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/include/c++/5.4.1
```
Comment (dagar) [inline cmake/nuttx/px4_impl_nuttx.cmake]: fix the path

**Comment by dagar** [comment]
Looking good so far. How do we use this? What are the dependencies?

Once it's working we'll need to add it to the build system or it's unlikely to continue working.

@LorenzMeier are those stack increases tolerable to support this?

**Comment by davids5** [comment]
@dagar - 

What is the value add here?

Who is going to vet the 24 builds on HW?

**Comment by dagar** [comment]
@davids5 it's not switching to clang, or building all firmware under clang, do doing anything unless it won't impact the regular builds.

The goal is adding the ability to optionally use clang if a developer is interested for various reasons. They would clearly be in "unsupported" territory. 

If we can get it working optionally I'd add a single build just to keep it working.

File: cmake/nuttx/px4_impl_nuttx.cmake (hunk 1 above)
Comment (mayanez) [inline cmake/nuttx/px4_impl_nuttx.cmake]: I'm not sure what the best way to get this path using CMake so I'm open to any suggestions.

File: cmake/common/px4_base.cmake (hunk 2 above)
Comment (mayanez) [inline cmake/common/px4_base.cmake]: I was having some problems at one point, but I'll see if I can enable it.

**Comment by mayanez** [comment]
Correct, its not for building all firmware under clang. It's more for development as I know vetting all 24 builds is going to be quite difficult. Having the ability to compile with Clang I think will open the platform up for additional research. For example, I'll be using PX4 as a development platform for testing various LLVM passes which I've written.

**Comment by mayanez** [comment]
I've added a Python script to resolve the hardcoded path issue.

With regards to the `-fcolor-diagnostics` flag the issue seems to be caused by clang using the g++ for linking which does not support the flag. I'll try to debug this further, but the rest of the code is working fine.

 I'd be more than happy to add support to the build system if possible.

**Comment by dagar** [comment]
Great, happy to see you found a solution for the paths. So how does a developer specify to build nuttx with clang? CC and CXX?

We should make sure the default builds are unaffected, add a new clang build, and get CI fully passing. Then it should be in a good state to review and merge.

**Comment by dagar** [comment]
What are the full dependencies? We have build environments in various docker images. They're broken up to keep them small as travis-ci/semaphore/etc pull the entire thing every build. https://github.com/PX4/containers

**Comment by mayanez** [comment]
The dependencies are Clang + GCC arm toolchain. I thought the clang docker image had GCC, but I guess I was mistaken. Looking over that repo link looks like there isn't one that matches so I'll create one and submit a PR.

File: cmake/nuttx/px4_impl_nuttx.cmake [hunk 4]
```diff
@@ -582,10 +607,57 @@ function(px4_os_add_flags)
 			-march=armv7-m
 			)
 	endif()
-	list(APPEND c_flags ${cpu_flags})
-	list(APPEND cxx_flags ${cpu_flags})
+
+	list(APPEND gcc_c_flags ${cpu_flags})
+
+	if (${CMAKE_C_COMPILER_ID} MATCHES ".*Clang.*" OR ${CMAKE_CXX_COMPILER_ID} MATCHES ".*Clang.*")
```
Comment (dagar) [inline cmake/nuttx/px4_impl_nuttx.cmake]: @mayanez could this be contained in Toolchain-arm-none-eabi-clang.cmake?

**Comment by dagar** [comment]
Is this PR completely replaced by https://github.com/PX4/Firmware/pull/6892?

**Pull request closed without merging**